system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1302,355 @@ handle_run_triggers (FlatpakSystemHelper   *object,
+          break;
+        }
+
+      if (file_info == NULL)
+        break;
+
+      name = g_file_info_get_name (file_info);
+      if (g_file_info_get_file_type (file_info) == G_FILE_TYPE_DIRECTORY &&
+          g_str_has_prefix (name, prefix))
+        {
+          g_autoptr(GFile) cache_dir = g_file_get_child (repo_tmpfile, name);
+          OngoingPull *pull;
+          gboolean dir_in_use = FALSE;
+
+          /* If this cache directory is used by any other ongoing pull, we
+           * should not be re(using) it. */
+          G_LOCK (ongoing_pulls);
+          for (guint i = 0; i < ongoing_pulls->len; i++)
+            {
+              pull = g_ptr_array_index (ongoing_pulls, i);
+              if (g_strcmp0 (pull->src_dir, name) == 0)
+                {
+                  dir_in_use = TRUE;
+                  break;
+                }
+            }
+          G_UNLOCK (ongoing_pulls);
+
+           if (!dir_in_use && check_for_repo_containing_partial_pull (cache_dir, commit, out_location))
+            return TRUE;
+        }
+    }
+
+  return FALSE;
+}
+
+static gboolean
+handle_get_revokefs_fd (FlatpakSystemHelper   *object,
+                        GDBusMethodInvocation *invocation,
+                        GUnixFDList           *arg_fdlist,
+                        guint                  arg_flags,
+                        const gchar           *arg_ref,
+                        const gchar           *arg_commit,
+                        const gchar           *arg_installation)
+{
+  g_autoptr(FlatpakDir) system = NULL;
+  g_autoptr(GUnixFDList) fd_list = NULL;
+  g_autoptr(GError) error = NULL;
+  g_auto(GStrv) parts = NULL;
+  g_autofree gchar *src = NULL;
+  g_autofree gchar *flatpak_dir = NULL;
+  g_autofree gchar *repo_tmp = NULL;
+  g_autofree gchar *cache_dir_basename = NULL;
+  g_autofree gchar *partial_pull_repo = NULL;
+  struct passwd *passwd;
+  OngoingPull *new_pull;
+  uid_t uid;
+  int fd_index = -1;
+  int exit_status= -1;
+
+  g_debug ("GetRevokefsFd %u %s %s %s", arg_flags, arg_ref, arg_commit, arg_installation);
+
+  system = dir_get_system (arg_installation, get_sender_pid (invocation), &error);
+  if (system == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  if (!check_revokefs_requirements (&error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  if (!get_connection_uid (invocation, &uid, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+  repo_tmp = g_build_filename (flatpak_dir, "repo", "tmp", NULL);
+  parts = flatpak_decompose_ref (arg_ref, NULL);
+
+   if (check_partial_pull_availability (repo_tmp, parts[1], arg_commit, &partial_pull_repo))
+    {
+      g_debug ("Found a partial pull for ref %s at: %s", parts[1], partial_pull_repo);
```

Comment:
It would perhaps be nice to check that this partial pull directory is owned by the correct user before deciding to use it.

Reply:
hmm, this might need a bit of thinking. The "correct user" is basically the `flatpak` user. All ongoing/partial pulls are owned by `flatpak` user. By "correct user", do you mean the (non-privileged)user who initiated the pull?

Reply:
I mean the `flatpak` user.

Reply:
> All ongoing/partial pulls are owned by `flatpak` user. 

I am saying that something should check or enforce this.

Reply:
We do this now in the `check_for_repo_containing_partial_pull` function itself.
See [here](https://github.com/flatpak/flatpak/pull/2657/commits/17a7f9b001a0c3898c9f80bd958e6bc4d4a22ee0#diff-9885ca1a8cc5c8690061c7fed45ddf20R1459)